File: frontend/src/services/api/portal.ts
import patientClient from "./patientClient";

export interface PRemedy { name: string; potency?: string; dosage?: string; frequency?: string; duration?: string; }
export interface PPrescription {
  id: string; remedies: PRemedy[]; dietary_advice?: string; lifestyle_advice?: string;
  precautions?: string; created_at?: string;
}
export interface PNotification { type: string; message: string; at?: string | null; }
export interface PDashboard {
  patient_name: string; recovery_status: string; last_visit_at?: string | null;
  next_followup_at?: string | null; current_prescription?: PPrescription | null; notifications: PNotification[];
}
export interface PFollowUp {
  id: string; followup_type: string; scheduled_at: string; responded: boolean;
  outcome?: string | null; wellness?: number | null;
}
export interface PTimelineEvent { type: string; label: string; at: string | null; meta?: any; }

export const portalAuth = {
  login: (phone: string, access_code: string) =>
    patientClient.post<{ access_token: string }>("/auth/login", { phone, access_code }),
  me: () => patientClient.get<{ id: string; full_name: string; phone: string; gender: string; language_pref?: string }>("/me"),
};

export const portalApi = {
  dashboard: () => patientClient.get<PDashboard>("/dashboard"),
  prescriptions: () => patientClient.get<PPrescription[]>("/prescriptions"),
  timeline: () => patientClient.get<{ events: PTimelineEvent[] }>("/timeline"),
  followups: () => patientClient.get<PFollowUp[]>("/followups"),
  respond: (id: string, data: { status: string; wellness?: number; notes?: string }) =>
    patientClient.post<PFollowUp>(`/followups/${id}/respond`, data),
  pdfUrl: (id: string) => `/api/v1/portal/prescriptions/${id}/pdf`,
};

File: backend/app/schemas/portal.py
from pydantic import BaseModel, Field
from typing import Optional, List, Any
from datetime import datetime
from uuid import UUID


class PatientLogin(BaseModel):
    phone: str
    access_code: str


class PatientToken(BaseModel):
    access_token: str
    token_type: str = "bearer"


class PatientProfile(BaseModel):
    id: UUID
    full_name: str
    phone: str
    gender: str
    language_pref: str = "en"

    class Config:
        from_attributes = True


class PatientPrescription(BaseModel):
    id: UUID
    remedies: List[Any]
    dietary_advice: Optional[str] = None
    lifestyle_advice: Optional[str] = None
    precautions: Optional[str] = None
    created_at: Optional[datetime] = None


class PatientFollowUp(BaseModel):
    id: UUID
    followup_type: str
    scheduled_at: datetime
    responded: bool
    outcome: Optional[str] = None
    wellness: Optional[int] = None


class FollowUpSubmit(BaseModel):
    # Patient-friendly: better / same / worse maps to the clinical outcome enum.
    status: str = Field(pattern="^(better|same|worse)$")
    wellness: Optional[int] = Field(default=None, ge=1, le=10)
    notes: Optional[str] = None


class Notification(BaseModel):
    type: str
    message: str
    at: Optional[datetime] = None


class PatientDashboard(BaseModel):
    patient_name: str
    recovery_status: str                     # friendly, non-clinical
    last_visit_at: Optional[datetime] = None
    next_followup_at: Optional[datetime] = None
    current_prescription: Optional[PatientPrescription] = None
    notifications: List[Notification] = []
